### ISSUE TITLE
[Question] req/rep inverted async demo

### ISSUE THREAD

**Issue by 0vercl0k** [nanomsg/nng#1302]
Hello,

I'm filing this issue to have an answer for a question that I have: I have a very similar need than the one described in the `async` demo. The only slight difference is that my server is actually a REQ and my clients are REPs.

I've first prototyped this in my project and realized that the library was behaving weird so I figured I'd try to make something smaller. I modified the `async` demo to do just that: make the server a REQ instead of a REP; instead of receiving then sending, it sends, then receives. Also modified the client to accommodate that.

To give a few more details on what I'm seeing (I haven't really debugged nor tried to find where / why this is happening yet): if I have X `aio` waiting on a `recv` (to be able to handle X simultaneous clients) and a client comes, the response of the server gets sent more than once; once with the expected `nng_ctx` but I can also see the same data getting sent with other `nng_ctx`s as well (I simply display the context identifier).

Before I write a reproducer, is the above expected to work, or is there some sort of limitation / something I misunderstood along the way regarding the APIs/REQ/REP? FWIW I am doing my tests with the latest release which is `v1.3.2`; and I've verified that the same things happen on both Linux / Windows with the `tcp` transport.

Cheers

**Comment by gdamore** [comment]
It should work, and this workflow is quite common.  My instinct is that there is something wrong in your modified demo.

**Comment by 0vercl0k** [comment]
Cool makes sense. And yeah I'm also thinking I'm getting it wrong somehow. Let me cook you a reproducer..

Cheers

**Comment by 0vercl0k** [comment]
All right so this is my reproducer file which is literally the `async` demo modified for having the server REQ and the client REP:

```c
// $ clang-11 repro.c -o repro -I ~/nng/include ~/libnng.a -lpthread -fsanitize=address
// $ ./repro server tcp://localhost:31337/
// [1] Sending 11223344..
// [2] Sending 11223344..
// [3] Sending 11223344..
// [1] Sent!
// [2] Sent!
// [3] Sent!
// [1] Received 4 bytes: 31337
//
// $ ./repro client tcp://localhost:31337/
// Received 4 bytes: 809609984
// Sent: 31337
//
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include <nng/nng.h>
#include <nng/protocol/reqrep0/req.h>
#include <nng/protocol/reqrep0/rep.h>
#include <nng/supplemental/util/platform.h>

#define PARALLEL 3

struct work {
    enum { INIT, SENT, RECVD } state;
    nng_aio *aio;
    nng_msg *msg;
    nng_ctx  ctx;
};

void
fatal(const char *func, int rv)
{
    fprintf(stderr, "%s: %s\n", func, nng_strerror(rv));
    exit(1);
}

void
server_cb(void *arg)
{
    struct work *work = arg;
    int rv;
    uint32_t magic;

    switch (work->state) {
    case INIT:
        if((rv = nng_msg_alloc(&work->msg, 0)) != 0) {
            fatal("nng_msg_alloc", rv);
        }

        uint32_t value = 11223344;//rand();
        if((rv = nng_msg_append_u32(work->msg, value) != 0)) {
            fatal("nng_msg_append_u32", rv);
        }

        printf("[%u] Sending %u..\n", work->ctx.id, value);
        nng_aio_set_msg(work->aio, work->msg);
        nng_ctx_send(work->ctx, work->aio);
        work->state = SENT;
        break;

    case SENT:
        if ((rv = nng_aio_result(work->aio)) != 0) {
            fatal("nng_aio_result", rv);
        }

        printf("[%u] Sent!\n", work->ctx.id);
        work->msg = NULL;
        work->state = RECVD;
        nng_ctx_recv(work->ctx, work->aio);
        break;

    case RECVD:
        if ((rv = nng_aio_result(work->aio)) != 0) {
            fatal("nng_ctx_recv", rv);
        }

        nng_msg *msg = nng_aio_get_msg(work->aio);
        printf("[%u] Received %zd bytes: %d\n", work->ctx.id, nng_msg_len(msg), *(uint32_t*)nng_msg_body(msg));
        work->state = INIT;
        nng_msg_free(msg);
        break;

    default:
        fatal("bad state!", NNG_ESTATE);
        break;
    }
}

struct work *
alloc_work(nng_socket sock)
{
    struct work *w;
    int          rv;

    if ((w = nng_alloc(sizeof(*w))) == NULL) {
        fatal("nng_alloc", NNG_ENOMEM);
    }

    if ((rv = nng_aio_alloc(&w->aio, server_cb, w)) != 0) {
        fatal("nng_aio_alloc", rv);
    }

    if ((rv = nng_ctx_open(&w->ctx, sock)) != 0) {
        fatal("nng_ctx_open", rv);
    }

    w->state = INIT;
    return (w);
}

int
server(const char *url)
{
    nng_socket   sock;
    struct work *works[PARALLEL];
    int          rv;
    int          i;

    rv = nng_req_open(&sock);
    if (rv != 0) {
        fatal("nng_req_open", rv);
    }

    for (i = 0; i < PARALLEL; i++) {
        works[i] = alloc_work(sock);
    }

    if ((rv = nng_listen(sock, url, NULL, 0)) != 0) {
        fatal("nng_listen", rv);
    }

    for (i = 0; i < PARALLEL; i++) {
        server_cb(works[i]);
    }

    for (;;) {
        nng_msleep(3600000);
    }
}

int
client(const char *url)
{
    nng_socket sock;
    int        rv;
    nng_msg *  msg;

    if ((rv = nng_rep_open(&sock)) != 0) {
        fatal("nng_rep_open", rv);
    }

    if ((rv = nng_dial(sock, url, NULL, 0)) != 0) {
        fatal("nng_dial", rv);
    }

    char *buffer = NULL;
    size_t buffer_size = 0;
    if ((rv = nng_recv(sock, &buffer, &buffer_size, NNG_FLAG_ALLOC)) != 0) {
        fatal("nng_recv", rv);
    }

    printf("Received %zd bytes: %u\n", buffer_size, *(uint32_t*)buffer);
    nng_free(buffer, buffer_size);
    buffer = NULL;
    buffer_size = 0;

    uint32_t magic = 31337;
    if ((rv = nng_send(sock, (void*)&magic, sizeof(magic), 0)) != 0) {
        fatal("nng_send", rv);
    }

    printf("Sent: %d\n", magic);
    nng_close(sock);
    return (0);
}

int
main(int argc, char **argv)
{
    int rc;
    if(argc == 3 && strcmp(argv[1], "server") == 0) {
        rc = server(argv[2]);
    }

    if(argc == 3 && strcmp(argv[1], "client") == 0) {
        rc = client(argv[2]);
    }

    exit(rc == 0 ? EXIT_SUCCESS : EXIT_FAILURE);
}
```

If I have one server and one client this is what I get:

```
$ ./repro server tcp://localhost:31337/
[1] Sending 11223344..
[2] Sending 11223344..
[3] Sending 11223344..
[1] Sent!
[2] Sent!
[3] Sent!
[1] Received 4 bytes: 31337
```

What I was expecting was a single `Sent!` and not as many as `PARALLEL`; what am I missing?

Cheers

PS: Also, I've noticed this violation with `fsanitize=memory`:

```
$ ./repro server tcp://localhost:31337/Uninitialized bytes in __interceptor_strcmp at offset 0 inside [0x701000000090, 1)
==3640==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x4b2c3e in nni_url_parse (repro+0x4b2c3e)
    #1 0x4a69f0 in nni_listener_create (repro+0x4a69f0)
    #2 0x49b1e9 in nng_listen (repro+0x49b1e9)
    #3 0x499139 in server (repro+0x499139)
    #4 0x499bcd in main (repro+0x499bcd)
    #5 0x7f37b2e0e0b2 in __libc_start_main
    #6 0x41d4ed in _start (repro+0x41d4ed)

SUMMARY: MemorySanitizer: use-of-uninitialized-value (repro+0x4b2c3e) in nni_url_parse
Exiting
```

**Comment by gdamore** [comment]
So you broke the state model entirely. The server always receives first, and then sends. What I think you should have done instead of this change was modify the demo so that the server *dials* and the client *listens*. I.e. instead of changing REQ to mean server, you would have the server still be REP, but make it *dial*.  The client will still be REQ, but it will *listen*. The key element here is don’t confuse REQ/REP (client/server) with the TCP (dialer/listener) relationship.  They are orthogonal; the REP socket is still the server socket regardless if it is listening for incoming connections or initiating outbound ones itself.  (In fact it’s even possible to have a server or that does *both* listen and dial.  I don’t recommend that though, just from a matter of complexity management. 😊 Sent from Mail for Windows 10 From: Axel SouchetSent: Thursday, October 15, 2020 6:14 PMTo: nanomsg/nngCc: gdamore; CommentSubject: Re: [nanomsg/nng] [Question] req/rep inverted async demo (#1302) All right so this is my reproducer file which is literally the async demo modified for having the server REQ and the client REP:// $ clang-11 repro.c -o repro -I ~/nng/include ~/libnng.a -lpthread -fsanitize=address// $ ./repro server tcp://localhost:31337/// [1] Sending 11223344..// [2] Sending 11223344..// [3] Sending 11223344..// [1] Sent!// [2] Sent!// [3] Sent!// [1] Received 4 bytes: 31337//// $ ./repro client tcp://localhost:31337/// Received 4 bytes: 809609984// Sent: 31337//#include <stdio.h>#include <stdlib.h>#include <string.h>#include <time.h> #include <nng/nng.h>#include <nng/protocol/reqrep0/req.h>#include <nng/protocol/reqrep0/rep.h>#include <nng/supplemental/util/platform.h> #define PARALLEL 3 struct work {    enum { INIT, SENT, RECVD } state;    nng_aio *aio;    nng_msg *msg;    nng_ctx  ctx;}; voidfatal(const char *func, int rv){    fprintf(stderr, "%s: %s\n", func, nng_strerror(rv));    exit(1);} voidserver_cb(void *arg){    struct work *work = arg;    int rv;    uint32_t magic;     switch (work->state) {    case INIT:        if((rv = nng_msg_alloc(&work->msg, 0)) != 0) {            fatal("nng_msg_alloc", rv);        }         uint32_t value = 11223344;//rand();        if((rv = nng_msg_append_u32(work->msg, value) != 0)) {            fatal("nng_msg_append_u32", rv);        }         printf("[%u] Sending %u..\n", work->ctx.id, value);        nng_aio_set_msg(work->aio, work->msg);        nng_ctx_send(work->ctx, work->aio);        work->state = SENT;        break;     case SENT:        if ((rv = nng_aio_result(work->aio)) != 0) {            fatal("nng_aio_result", rv);        }         printf("[%u] Sent!\n", work->ctx.id);        work->msg = NULL;        work->state = RECVD;        nng_ctx_recv(work->ctx, work->aio);        break;     case RECVD:        if ((rv = nng_aio_result(work->aio)) != 0) {            fatal("nng_ctx_recv", rv);        }         nng_msg *msg = nng_aio_get_msg(work->aio);        printf("[%u] Received %zd bytes: %d\n", work->ctx.id, nng_msg_len(msg), *(uint32_t*)nng_msg_body(msg));        work->state = INIT;        nng_msg_free(msg);        break;     default:        fatal("bad state!", NNG_ESTATE);        break;    }} struct work *alloc_work(nng_socket sock){    struct work *w;    int          rv;     if ((w = nng_alloc(sizeof(*w))) == NULL) {        fatal("nng_alloc", NNG_ENOMEM);    }     if ((rv = nng_aio_alloc(&w->aio, server_cb, w)) != 0) {        fatal("nng_aio_alloc", rv);    }     if ((rv = nng_ctx_open(&w->ctx, sock)) != 0) {        fatal("nng_ctx_open", rv);    }     w->state = INIT;    return (w);} intserver(const char *url){    nng_socket   sock;    struct work *works[PARALLEL];    int          rv;    int          i;     rv = nng_req_open(&sock);    if (rv != 0) {        fatal("nng_req_open", rv);    }     for (i = 0; i < PARALLEL; i++) {        works[i] = alloc_work(sock);    }     if ((rv = nng_listen(sock, url, NULL, 0)) != 0) {        fatal("nng_listen", rv);    }     for (i = 0; i < PARALLEL; i++) {        server_cb(works[i]);    }     for (;;) {        nng_msleep(3600000);    }} intclient(const char *url){    nng_socket sock;    int        rv;    nng_msg *  msg;     if ((rv = nng_rep_open(&sock)) != 0) {        fatal("nng_rep_open", rv);    }     if ((rv = nng_dial(sock, url, NULL, 0)) != 0) {        fatal("nng_dial", rv);    }     char *buffer = NULL;    size_t buffer_size = 0;    if ((rv = nng_recv(sock, &buffer, &buffer_size, NNG_FLAG_ALLOC)) != 0) {        fatal("nng_recv", rv);    }     printf("Received %zd bytes: %u\n", buffer_size, *(uint32_t*)buffer);    nng_free(buffer, buffer_size);    buffer = NULL;    buffer_size = 0;     uint32_t magic = 31337;    if ((rv = nng_send(sock, (void*)&magic, sizeof(magic), 0)) != 0) {        fatal("nng_send", rv);    }     printf("Sent: %d\n", magic);    nng_close(sock);    return (0);} intmain(int argc, char **argv){    int rc;    if(argc == 3 && strcmp(argv[1], "server") == 0) {        rc = server(argv[2]);    }     if(argc == 3 && strcmp(argv[1], "client") == 0) {        rc = client(argv[2]);    }     exit(rc == 0 ? EXIT_SUCCESS : EXIT_FAILURE);}If I have one server and one client this is what I get:$ ./repro server tcp://localhost:31337/[1] Sending 11223344..[2] Sending 11223344..[3] Sending 11223344..[1] Sent![2] Sent![3] Sent![1] Received 4 bytes: 31337What I was expecting was a single Sent! and not as many as PARALLEL; what am I missing?CheersPS: Also, I've noticed this violation with fsanitize=memory:$ ./repro server tcp://localhost:31337/Uninitialized bytes in __interceptor_strcmp at offset 0 inside [0x701000000090, 1)==3640==WARNING: MemorySanitizer: use-of-uninitialized-value    #0 0x4b2c3e in nni_url_parse (repro+0x4b2c3e)    #1 0x4a69f0 in nni_listener_create (repro+0x4a69f0)    #2 0x49b1e9 in nng_listen (repro+0x49b1e9)    #3 0x499139 in server (repro+0x499139)    #4 0x499bcd in main (repro+0x499bcd)    #5 0x7f37b2e0e0b2 in __libc_start_main    #6 0x41d4ed in _start (repro+0x41d4ed) SUMMARY: MemorySanitizer: use-of-uninitialized-value (repro+0x4b2c3e) in nni_url_parseExiting—You are receiving this because you commented.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by 0vercl0k** [comment]
I think there was something wrong with the reply - the layout is all messed up on Github ^^'; I'll try to extract the interesting bits and do the changes you suggested 👍🏼 

Thanks for the fast turnaround!

Cheers

**Comment by 0vercl0k** [comment]
OK so I digested your last answer a bit, as well as did a bit of testing to see how things behave. So I've done the change you suggested which was to simply swap who listens and who dials right? This means my client is now listening and the server connects to it. Good this does what I expect to do.

The caveat with this is that in my architecture I have several concurrent clients and so I run into issues when they try to all listen at the same time (reuse address; I reckon that this is probably solvable). For my tests both the server and clients run on localhost but assuming clients run a bit everywhere across a fleet of machine then where does the server now dials to? That's basically the reason why I swapped REQ/REP instead of changing who listens and who dials. But again, I feel I might be misunderstanding something.

Cheers

**Comment by gdamore** [comment]
What you have is a common problem -- you want to have a few number of addresses (just 1 or 2) for things to dial. 

The solution is to use nng_device() to create a proxy, which listens on two addresses one is for REQ, and the other is for REP.

Then both servers and clients can dial, and can dynamically come and go.

**Comment by 0vercl0k** [comment]
Hmmm ok. It is starting to sound like nng might not be the solution I was
looking for for my project then.

It seems like it's turning something that I can easily do with a little
win/linux sockets layer into something a bit more involved with an extra
layer that I don't think I need.

Cheers

**Comment by gdamore** [comment]
Depends on what you need.  Actually building what I *think* you want is super trivial with nng.

The proxy (which is a kind of broker in the middle that lets multiple clients and servers rendezvous with each other) is simply just open two sockets (one REQ, one REP), call nng_listen() on the appropriate address, then pass the sockets to nng_device().

Then your clients or works can just use a socket along with nng_dial() to the appropriate broker address.

(Btw, you can prototype this without a broker -- from the client or the server the broker looks the same as a client or server.)

If you don't want a broker,  to share the address, you can use multiple clients at *different* addresses.

I suppose at some point I should just start supporting SO_REUSEPORT which would allow you to have multiple clients at the *same address*.  (Admittedly I've not done this already for two reasons: 1. SO_REUSEPORT isn't fully portable, and 2. SO_REUSEPORT breaks the normal TCP security model.)

**Comment by 0vercl0k** [comment]
What I need is pretty simple; I have a single server that needs to
distribute work to clients; clients compute a result and send it back to
the server. The server aggregates the results and is the only one having
the bigger picture. I don't expect ever going over 100 clients and I'll
mostly usually have ten-ish to give you an idea. That's also why I'd like
to KISS. Basically, with traditional sockets I would have the server
listen, accept & poll.

What you are describing is basically the architecture made in one of the
box in this illustration right https://250bpm.com/blog:14/prio2.png?

Cheers

**Comment by gdamore** [comment]
Yep but only one half of that.  You don't need two brokers.  

But based on what you're telling me maybe you can get by with just the the workers dialing the distributor.  The clients are REP but they dial the server.  This is dead easy.  I thought you needed to have multiple servers.

**Comment by gdamore** [comment]
So one "server" that is really a REQ client but calls listen. 

The clients are REP workers that dial the server. 

The server socket will automatically load balance to the clients.

**Comment by 0vercl0k** [comment]
All right, let me try this then!

Cheers

**Comment by 0vercl0k** [comment]
So it kinda feels like I'm going full circle at this point - isn't your suggestion what I had done with the `repro.c` I pasted earlier? 🤔

**Comment by gdamore** [comment]
The *protocol* state model is independent of the listen/dial relationship.

Hmm... I might have fallen into my own trap of misunderstand what is meant (in your code) by server vs. client.

To me, REP is *always* the server, regardless of the TCP dialer/listener relationship.  Going back and looking at your repro code, I see instead you wanted to have the REQ be the "server".

I think I probably need to try your code, because on the face of it it seems like it should work, so I'm clearly missing something in the details.

**Comment by 0vercl0k** [comment]
Right - yeah no worries this gets confusing fast 🙂. Lemme know if I can help out.

Cheers

**Comment by gdamore** [comment]
Ok, I know what is happening.

The parallel REQ workers are each trying to send work.  When the worker (client) connects, they all actually do try to send the message.

The message is only delivered to a single socket -- because flow control pressures are applied to allow back-pressure.

When the first client comes, and *then* disconnects, we basically wind up waking the sender telling them there message was sent.  This is a lie.  It's been *queued* for send.  This wouldn't happen if the client was long lived.

This is also a bug really -- we should not mark this, but should ensure that the REQ socket exerts backpressure through the context until we actually have posted the message to a socket.

There's a deeper bug here, which is that somehow after servicing a couple of requests,  the server is supposed to keep moving forward.  That's not happening.  I wonder if this is a recent regression.

 - because the worker's TCP socket buffer has enough room to queue up all those requests.  Since there is no where else those requests could go (only one client), they all land there.

Unfortunately, your client worker is short lived, and so disconnects.  This causes those outstanding requests to suffer a callback.

**Comment by 0vercl0k** [comment]
Great to hear - I am not going mad after all.

Want me to try it out on previous versions?

Cheers

**Comment by gdamore** [comment]
I'm going to test some prior versions ... stay tuned.

**Comment by gdamore** [comment]
The bug goes back to at least 1.2.  Which is far enough that its not a recent regression.

**Comment by gdamore** [comment]
So, there isn't actually a bug really -- or at least the possible bug I'm worried about isn't a real concern.

Your program fails to move from the final state on the server to the INIT state -- you set the state, but there is no callback to reenter the init state.  If you call server_cb() from the RECVD state, that will cause you to reenter the INIT state handler.

Alternatively, you could call nng_msg_send() from that callback, and skip the INIT state.  I'd recommend that actually.

This is the hang that I was really worried about, and it turns out it is in the application code rather than the library.

I think we wind up completing the request because of TCP buffering -- under the hood the client accepts the next request for work (at the TCP level) but we disconnect the socket, and so the server has to resend.

The REQ/REP framework is not ideally designed for a REP server that swings by, does a request, and sends the response.  (There are issues with that too, besides, such as the possibility of losing the reply.)

The assumption is that clients will be long lived.

Anyway, once I fixed the application code to move to the next state after receiving the reply, everything seems to work.

**Comment by 0vercl0k** [comment]
Wow I see, embarrassing. Well, I'm sorry to have not seen that myself, really appreciate the help. FWIW the client are long-lived in my application.

Should I keep the issue open to track the "When the first client comes, and then disconnects, we basically wind up waking the sender telling them there message was sent. This is a lie. It's been queued for send. This wouldn't happen if the client was long lived." issue?

Cheers

**Comment by gdamore** [comment]
Let's leave this issue open for now.  I'll figure out what to do about it later.

**Comment by 0vercl0k** [comment]
So I've been keeping playing with this and I've hit another weird thing.

After having a server and a client run for a bit of a while I am noticing the server having `nng_ctx_receive` fail with `Incorrect state` -- is this something that is expected and that I should handle gracefully or it's a sign that something else might be wrong? 

If that's any help, I've noticed that this happens always when two different `aio`s ends up sending which again is odd as I have a single client:

```
...
[3] Received 4 bytes: 92248
[3] Sending 318324966..
[3] Sent!
[4] Received 4 bytes: 92249
[4] Sending 174542394..
[4] Sent!
[2] Received 4 bytes: 92250
[2] Sending 724241990..
[2] Sent!
[6] Received 4 bytes: 92251
[6] Sending 1915554257..
[6] Sent!
[7] Received 4 bytes: 92252
[7] Sending 734673185..
[7] Sent!
[9] Received 4 bytes: 92253
[9] Sending 1209026776..
[9] Sent!
[8] Received 4 bytes: 92254
[8] Sending 557867750..
[8] Sent!
[10] Received 4 bytes: 92255
[10] Sending 1688156489..
[10] Sent!
[5] Received 4 bytes: 92256
[5] Sending 1555917759..
[5] Sent!
[1] Sent!
nng_ctx_recv: Incorrect state
```

Cheers

**Comment by gdamore** [comment]
for REQ, you cannot call recv unless you have first sent a message.  You should not use the same context with multiple AIOs, as the context only keeps track of it's state -- i.e. the state machine is part of the context, not part of the AIO.

IF you call send, and send again, then you should only call recv once which will wait for the answer for the most recent send.  (Calling send the second time cancels the first send.)

**Comment by 0vercl0k** [comment]
Right - I mean I am triggering the above with the same code in this issue
(after fixing the code so that it goes back to the 'INIT' state again)
which means that there is a context per AIOs and they obey the sequencing
of recv/send.

Cheers

**Comment by gdamore** [comment]
Can you post your fixed code, because this doesn't make sense to me as to why it would happen.

**Comment by 0vercl0k** [comment]
The new code is not much different than the previous one:

- I've fixed the state issue that you identified in my previous proof-of-concept and fixed it with the (not recommended) way by doing a reentrant call,

- The client side now simulates a long lived client that blasts 1M:

```c
// $ clang-11 repro.c -o repro -I ~/nng/include ~/libnng.a -lpthread -fsanitize=address
// $ ./repro server tcp://localhost:31337/
// [1] Sending 11223344..
// [2] Sending 11223344..
// [3] Sending 11223344..
// [1] Sent!
// [2] Sent!
// [3] Sent!
// [1] Received 4 bytes: 31337
//
// $ ./repro client tcp://localhost:31337/
// Received 4 bytes: 809609984
// Sent: 31337
//
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include <nng/nng.h>
#include <nng/protocol/reqrep0/req.h>
#include <nng/protocol/reqrep0/rep.h>
#include <nng/supplemental/util/platform.h>

#define PARALLEL 10

struct work {
    enum { INIT, SENT, RECVD } state;
    nng_aio *aio;
    nng_msg *msg;
    nng_ctx  ctx;
};

void
fatal(const char *func, int rv)
{
    fprintf(stderr, "%s: %s\n", func, nng_strerror(rv));
    exit(1);
}

void
server_cb(void *arg)
{
    struct work *work = arg;
    int rv;
    uint32_t magic;
    switch (work->state) {
    case INIT:
        if((rv = nng_msg_alloc(&work->msg, 0)) != 0) {
            fatal("nng_msg_alloc", rv);
        }

        uint32_t value = rand();
        if((rv = nng_msg_append(work->msg, &value, sizeof(value)) != 0)) {
            fatal("nng_msg_append_u32", rv);
        }

        printf("[%u] Sending %u..\n", work->ctx.id, value);
        nng_aio_set_msg(work->aio, work->msg);
        nng_ctx_send(work->ctx, work->aio);
        work->state = SENT;
        break;

    case SENT:
        if ((rv = nng_aio_result(work->aio)) != 0) {
            fatal("nng_aio_result", rv);
        }

        work->msg = NULL;
        printf("[%u] Sent!\n", work->ctx.id);
        work->state = RECVD;
        nng_ctx_recv(work->ctx, work->aio);
        break;

    case RECVD:
        if ((rv = nng_aio_result(work->aio)) != 0) {
            fatal("nng_ctx_recv", rv);
        }

        nng_msg *msg = nng_aio_get_msg(work->aio);
        printf("[%u] Received %zd bytes: %d\n", work->ctx.id, nng_msg_len(msg), *(uint32_t*)nng_msg_body(msg));
        work->state = INIT;
        nng_msg_free(msg);
        server_cb(arg);
	break;

    default:
        fatal("bad state!", NNG_ESTATE);
        break;
    }
}

struct work *
alloc_work(nng_socket sock)
{
    struct work *w;
    int          rv;

    if ((w = nng_alloc(sizeof(*w))) == NULL) {
        fatal("nng_alloc", NNG_ENOMEM);
    }

    if ((rv = nng_aio_alloc(&w->aio, server_cb, w)) != 0) {
        fatal("nng_aio_alloc", rv);
    }

    if ((rv = nng_ctx_open(&w->ctx, sock)) != 0) {
        fatal("nng_ctx_open", rv);
    }

    w->state = INIT;
    return (w);
}

int
server(const char *url)
{
    nng_socket   sock;
    struct work *works[PARALLEL];
    int          rv;
    int          i;

    rv = nng_req_open(&sock);
    if (rv != 0) {
        fatal("nng_req_open", rv);
    }

    for (i = 0; i < PARALLEL; i++) {
        works[i] = alloc_work(sock);
    }

    if ((rv = nng_listen(sock, url, NULL, 0)) != 0) {
        fatal("nng_listen", rv);
    }

    for (i = 0; i < PARALLEL; i++) {
        server_cb(works[i]);
    }

    for (;;) {
        nng_msleep(3600000);
    }
}

int
client(const char *url)
{
    nng_socket sock;
    int        rv;
    nng_msg *  msg;

    if ((rv = nng_rep_open(&sock)) != 0) {
        fatal("nng_rep_open", rv);
    }

    if ((rv = nng_dial(sock, url, NULL, 0)) != 0) {
        fatal("nng_dial", rv);
    }

    for(uint32_t i = 0; i < 1000000; i++) {
        char *buffer = NULL;
        size_t buffer_size = 0;
        if ((rv = nng_recv(sock, &buffer, &buffer_size, NNG_FLAG_ALLOC)) != 0) {
            fatal("nng_recv", rv);
        }

        printf("Received %zd bytes: %u\n", buffer_size, *(uint32_t*)buffer);
        nng_free(buffer, buffer_size);
        buffer = NULL;
        buffer_size = 0;

        if ((rv = nng_send(sock, (void*)&i, sizeof(i), 0)) != 0) {
            fatal("nng_send", rv);
        }

        printf("Sent: %d\n", i);
    }
    nng_close(sock);
    return (0);
}

int
main(int argc, char **argv)
{
    int rc;
    if(argc == 3 && strcmp(argv[1], "server") == 0) {
        rc = server(argv[2]);
    }

    if(argc == 3 && strcmp(argv[1], "client") == 0) {
        rc = client(argv[2]);
    }

    exit(rc == 0 ? EXIT_SUCCESS : EXIT_FAILURE);
}
```

I've reproduced the above on Linux with one server and one client run with `./repro server tcp://localhost:31337/` / `./repro client tcp://localhost:31337/`.

Cheers

**Comment by gdamore** [comment]
Ok, I've tracked this down... the problem is a race condition in your code caused by an out of order operation.

You have the following:

```
        printf("[%u] Sending %u..\n", work->ctx.id, value);
        nng_aio_set_msg(work->aio, work->msg);
        nng_ctx_send(work->ctx, work->aio);
        work->state = SENT;
```

The problem is that after you send the work, we can wind up having the callback execute *before* the next line (because everything is asynch), which means you state machine will be busted and you wind up calling send *twice*.

The solution is trivial.

Move the assignment to work->state *before* you start the operation with nng_ctx_send().

I was able to confirm that this was both the source of your problem, and that my proposed fix works.

**Comment by 0vercl0k** [comment]
Wow I see it, that was sneaky; thanks for the help again. I'm going to play with this a bit more this week - would you be interested to contribute something like the above as part of a demo by the way? Least I can do at this point :).

Cheers

**Comment by gdamore** [comment]
I'm always open to contributions.
